File: frontend/src/api/readLater.ts
import api from './client';

export interface ReadLaterItem {
  id: string;
  user_id: string;
  title?: string;
  url: string;
  domain?: string;
  excerpt?: string;
  full_text?: string;
  cover_image?: string;
  status: 'unread' | 'reading' | 'read' | 'archived';
  is_favorite: boolean;
  read_progress: number;
  source: string;
  item_status: string;
  knowledge_id?: string;
  created_at: string;
  updated_at: string;
}

export interface ReadLaterCreateData {
  url: string;
  title?: string;
  excerpt?: string;
  source?: string;
}

export interface ReadLaterUpdateData {
  title?: string;
  excerpt?: string;
  status?: 'unread' | 'reading' | 'read' | 'archived';
  is_favorite?: boolean;
  read_progress?: number;
}

export interface ReadLaterFilters {
  status?: string;
  is_favorite?: boolean;
  q?: string;
  skip?: number;
  limit?: number;
}

export const readLaterApi = {
  list: (params?: ReadLaterFilters) => api.get<ReadLaterItem[]>('/api/v1/read-later/items', { params }),
  create: (data: ReadLaterCreateData) => api.post<ReadLaterItem>('/api/v1/read-later/items', data),
  get: (id: string) => api.get<ReadLaterItem>(`/api/v1/read-later/items/${id}`),
  update: (id: string, data: ReadLaterUpdateData) => api.put<ReadLaterItem>(`/api/v1/read-later/items/${id}`, data),
  delete: (id: string) => api.delete(`/api/v1/read-later/items/${id}`),
  fetchContent: (id: string) => api.post<ReadLaterItem>(`/api/v1/read-later/items/${id}/fetch-content`),
  saveToKnowledge: (id: string, tagIds?: string[]) =>
    api.post<{ success: boolean; knowledge_id: string }>(`/api/v1/read-later/items/${id}/save-to-knowledge`, { tag_ids: tagIds }),
};

File: backend/app/schemas/read_later.py
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional, List


class ReadLaterCreate(BaseModel):
    url: str = Field(..., min_length=1, max_length=2048)
    title: Optional[str] = Field(None, max_length=500)
    excerpt: Optional[str] = Field(None, max_length=2000)
    source: Optional[str] = Field("manual", max_length=50)


class ReadLaterUpdate(BaseModel):
    title: Optional[str] = Field(None, max_length=500)
    excerpt: Optional[str] = Field(None, max_length=2000)
    status: Optional[str] = Field(None, pattern="^(unread|reading|read|archived)$")
    is_favorite: Optional[bool] = None
    read_progress: Optional[int] = Field(None, ge=0, le=100)


class ReadLaterResponse(BaseModel):
    id: str
    user_id: str
    title: Optional[str]
    url: str
    domain: Optional[str]
    excerpt: Optional[str]
    full_text: Optional[str]
    cover_image: Optional[str]
    status: str
    is_favorite: bool
    read_progress: int
    source: str
    item_status: str
    knowledge_id: Optional[str]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ReadLaterSaveToKnowledgeRequest(BaseModel):
    tag_ids: Optional[List[str]] = Field(None, description="Optional tag IDs or names to attach")
